[E2E][JF] Add skeleton and basic functionality for JF apps

doru91 opened the pull request:
Role of this PR is to introduce the two JF apps: jf-control-app and jf-admin-app

High-level architecture is described in [Slide 4](https://docs.google.com/presentation/d/1eOGmYHz6L1jsVTAbpDpZjaD-avM_KJ9L/edit#slide=id.p4) and the associated use-case (which will be fully implemented in following PRs) is in [Slides 5-7](https://docs.google.com/presentation/d/1eOGmYHz6L1jsVTAbpDpZjaD-avM_KJ9L/edit#slide=id.p5).

Associated SDK execution plan can be found in [project-chip/123](https://github.com/orgs/project-chip/projects/123/views/1), SDK section.

Fix https://github.com/project-chip/connectedhomeip/issues/38195, https://github.com/project-chip/connectedhomeip/issues/38179, https://github.com/project-chip/connectedhomeip/issues/38198: 

- PR is split in multiple commits with clear description for easier review
- jf-control-app is just a duplicate of chip-tool on top of which we are adding specific JF functionality (review of those files is not needed). This duplication is not something new, I also noticed that other features were doing the same for their tools (e.g.: [fabric-admin@FabricSync](https://github.com/project-chip/connectedhomeip/tree/master/examples/fabric-admin); [camera-controller](https://github.com/project-chip/connectedhomeip/tree/master/examples/camera-controller)) L.E.: As suggested by @chrisdecenzo found a middleground solution (similar to what tv-casting-app does): copied only the files that had to be JF-modified while referencing the untouched ones from their original location in examples/chip-tool.
- jfc plays the role of the commissioner while jfa plays the role of the JF Administrator
- jfa lacks the JF Administrator Cluster until https://github.com/project-chip/connectedhomeip/pull/34637 is merged
- more details about the architecture can be found in the guide introduced by this PR - [docs/guides/joint_fabric_guide.md](https://github.com/doru91/connectedhomeip/blob/feature/jf/docs/guides/joint_fabric_guide.md)

### Testing
Tested using the detailed instructions from docs/guides/joint_fabric_guide.md

doru91 committed:
[E2E][JF] jf-admin-app: initial version

Signed-off-by: Doru Gucea <doru-cristian.gucea@nxp.com>

doru91 committed:
[E2E][JF] jf-control-app: initial version

Signed-off-by: Doru Gucea <doru-cristian.gucea@nxp.com>

doru91 committed:
[E2E][JF] jf-admin-app: rpc: server: add joint fabric service

Add OwnershipTransfer function.

Signed-off-by: Doru Gucea <doru-cristian.gucea@nxp.com>

doru91 committed:
[E2E][JF] jf-control-app: add anchor flag for the pairing

First device commissioned by jfc has to be the Anchor Administrator, in which case
the '--anchor true' has to be specfied for the pairing command.

Once the Anchor Administrator has been commissioned, subsequent commissioning procedures
will use internally 'skip-commissioning-complete' parameter such that the commissioning
procedure is finalized by the Anchor Administrator.

Signed-off-by: Doru Gucea <doru-cristian.gucea@nxp.com>

doru91 committed:
[E2E][JF] jf-admin-app: finalize standard commissioning flow

CommissioningComplete fails for the moment because jfa lacks the required cats.

Signed-off-by: Doru Gucea <doru-cristian.gucea@nxp.com>

doru91 committed:
[core-SDK] ExampleOperationalCredentialsIssuer: add support for CaseAdminSubject

Add simple API to allow setting the caseAdminSubject.
One user of this API is JF (jf-control-app).

Signed-off-by: Doru Gucea <doru-cristian.gucea@nxp.com>

doru91 committed:
[E2E][JF] jf-control-app: issue correct CATs during standard commissioning

Additions:
- jfc receives a NOC with Administrator CAT at init time;
- jfa receives a NOC with Administrator CAT + Anchor CAT during commissioning
(added the check for CAT presence on jf-admin-app);
- CaseAdminSubject field of the AddNOC commands invoked by jfc contains
the Administrator CAT;
- initial value for both CATs is initially set to 1 and it's configurable from
CHIPProjectAppConfig.h

With this change in place, jfc executes only the commissioning steps then
ownership is transferred to jfa which completes the standard commissionning
procedure.

Signed-off-by: Doru Gucea <doru-cristian.gucea@nxp.com>

doru91 committed:
[core-SDK] CASEAuthTag.h: Add Admin and Anchor CAT simple API

Signed-off-by: Doru Gucea <doru-cristian.gucea@nxp.com>

doru91 committed:
Run restyled and spell-checker

Signed-off-by: Doru Gucea <doru-cristian.gucea@nxp.com>

doru91 committed:
[E2E][JF] jf-control-app/jf-admin-app: Fixes

* add initial JF guide/fix README
* add option for specifing the RPC server params on jf-control-app
* remove accidentally pushed files
* fix formatting to avoid ci/cv complaining
* enable ConfigurationVersion@BINFO to avoid ci/cv complaining
* core-SDK: add JointFabric logging namespace
* use custom credentials issuer (placed in the jfc folder)
* issue Anchor ICAC with "jf-anchor-icac" in its Subject DN
* apply suggestions from code review

Signed-off-by: Doru Gucea <doru-cristian.gucea@nxp.com>

bzbarsky-apple commented on src/lib/core/CASEAuthTag.h:
```diff
@@ -61,6 +63,16 @@ constexpr uint16_t GetCASEAuthTagVersion(CASEAuthTag aCAT)
     return static_cast<uint16_t>(aCAT & kTagVersionMask);
 }
 
+constexpr CASEAuthTag GetAdminCATHavingVersion(uint16_t version)
```
```suggestion
constexpr CASEAuthTag GetAdminCATWithVersion(uint16_t version)
```

doru91 replied:
Suggestion committed.

bzbarsky-apple commented on src/lib/core/CASEAuthTag.h:
```diff
@@ -61,6 +63,16 @@ constexpr uint16_t GetCASEAuthTagVersion(CASEAuthTag aCAT)
+    return ((static_cast<uint32_t>(kAdminCATIdentifier) << 16) | version);
+}
+
+constexpr CASEAuthTag GetAnchorCATHavingVersion(uint16_t version)
```
```suggestion
constexpr CASEAuthTag GetAnchorCATWithVersion(uint16_t version)
```

doru91 replied:
Suggestion committed.

chrisdecenzo commented on docs/guides/joint_fabric_guide.md:
```diff
@@ -0,0 +1,253 @@
+
+![matter_joint_fabric](images/matter_joint_fabric.png)
+
+jf-control-app acts as a commissioner and control plane for jf-admin-app. It
```
Might call out that jf-control-app is acting as the JF Anchor and so it too has the JF Administrator Cluster and the JF Datastore Cluster, right?

doru91 replied:
Nope: jf-admin-app acts as JF Anchor so it has an instance of the JF Administrator Cluster and of the JF Datastore Cluster. jf-control-app has an operational identify and a NOC with Administrator CAT but those are used only for the initial onboarding of the jf-admin-app. Details, links to relevant documents, architecture design can be found in the PR description, pls check.

chrisdecenzo commented on examples/jf-control-app/commands/clusters/ClusterCommand.h:
```diff
@@ -0,0 +1,285 @@
+/*
```
Are the files in commands/clusters, commands/common, commands/icd, etc actually unique to jf-control-app? 

I'm wondering if these can be included from a common location.

doru91 replied:
Please check the PR description: "jf-control-app is just a duplicate of chip-tool on top of which we are adding specific JF functionality (review of those files is not needed). This duplication is not something new, I also noticed that other features were doing the same for their tools (e.g.: fabric-admin@FabricSync; camera-controller)"

doru91 replied:
Thanks for the suggestion, in the end copied only the files that are unique to the jf-control-app and referenced the untouched files from the examples/chip-tool folder.

chrisdecenzo reviewed (approved):
please see comments

doru91 committed:
[E2E][JF] Refactor jf-control-app files

As also done by tv-casting-app, just reference the chip-tool files
which are untouched by JF functionality. Only copy and modify the
relevant files.

Signed-off-by: Doru Gucea <doru-cristian.gucea@nxp.com>
